### ISSUE TITLE
[Win] Text fields with org.eclipse.swt.internal.win32.Text.use_WS_BORDER = true have excentered caret

### ISSUE THREAD

**Issue by tmssngr** [eclipse-platform/eclipse.platform.swt#416]
**Describe the bug**
If the option org.eclipse.swt.internal.win32.Text.use_WS_BORDER is set to true, the caret is vertically excentered. Windows or SWT bug?

**To Reproduce**
Please run the following code on Windows 10:
```
import org.eclipse.swt.*;
import org.eclipse.swt.graphics.*;
import org.eclipse.swt.layout.*;
import org.eclipse.swt.widgets.*;

public class ThemeTextTest {

	public static void main(String[] args) {
		final Display display = new Display();
		display.addListener(SWT.Skin, event -> {
			if (event.widget instanceof Composite composite) {
				composite.setBackground(new Color(32, 32, 32));
				composite.setForeground(new Color(192, 192, 192));
			}
			else if (event.widget instanceof Text text) {
				text.setBackground(new Color(32, 32, 32));
				text.setForeground(new Color(192, 192, 192));
			}
		});
		display.setData("org.eclipse.swt.internal.win32.Text.use_WS_BORDER", true);

		final Shell shell = new Shell(display);
		shell.setLayout(new GridLayout(1, true));

		final Text text1 = new Text(shell, SWT.SINGLE | SWT.BORDER);
		text1.setLayoutData(new GridData(GridData.FILL, GridData.BEGINNING, true, false));

		final Text text2 = new Text(shell, SWT.SINGLE | SWT.BORDER | SWT.PASSWORD);
		text2.setLayoutData(new GridData(GridData.FILL, GridData.BEGINNING, true, false));

		shell.setSize(400, 200);
		shell.open();

		while (!shell.isDisposed()) {
			if (!display.readAndDispatch()) {
				display.sleep();
			}
		}

		display.dispose();
	}
}

```
You will notice that the caret sticks to the top of the control.

**Expected behavior**
The caret should be vertically centered. Just like when uncommenting the `display.setData` line.

**Screenshots**
With `org.eclipse.swt.internal.win32.Text.use_WS_BORDER = true`:
![nsborder](https://user-images.githubusercontent.com/6029561/193211867-cf7270eb-59cc-41ca-b9a4-4522cf43f013.png)
Default:
![default-border](https://user-images.githubusercontent.com/6029561/193211970-984aec6a-e6ef-43f5-8a8c-480e9f1e79a1.png)

**Environment:**
1. Select the platform(s) on which the behavior is seen:
- - [ ] All OS
- - [x] Windows
- - [ ] Linux
- - [ ] macOS 

**Workaround (or) Additional context**
Don't set this option and live with the high contrast border.

**Comment by SyntevoAlex** [comment]
The problem here is that Windows `Edit` control (used for SWT `Text`) does something weird:
* when created with `WS_BORDER` style, it removes this style, and remembers that border is needed
* border is then painted in client area (the area supposed to be painted by `SWT.Paint` etc).

The problem with said behavior is that `SWT.Paint` etc will now overpaint border, because they expect that borders are outside of client area.

In Bug 560316, I added a workaround to force Windows to use the usual approach for painting borders in Text:
https://github.com/syntevo/eclipse.platform.swt/blob/3cd2dc7cb792b9be2c78013774b23a7cbecd4560/bundles/org.eclipse.swt/Eclipse%20SWT/win32/org/eclipse/swt/widgets/Text.java#L313

But the downside of this workaround is that Windows paints text (and caret) 1px closer to the border, which results in the problem described here.

A better workaround would be to let Windows do what it wants, but in `SWT.Paint` etc, adjust GC coordinates so that coordinate (0,0) is after the border and there's no way to overpaint the border. Still, there will be a minor problem that `Control.getBounds()` will return bounds including border, but I think that this won't really bother anyone.

To summarize, a solution is possible, but it's not very trivial and not completely perfect.
